matlab_simulation/08-planning/shortestpath.m

```diff
@@ -20,71 +49,224 @@
+                dtogo = norm(nodes(neigh(i),:)-nodes(finish,:));
+                % If neighbour is not in open set, add it   
+                if (isempty(found))
+                   OpenSet = [OpenSet; neigh(i) bestnode(1) dtogo+dcur dcur];             
+                else
+                   if (dcur < OpenSet(found,4))
+                      OpenSet(found,:) = [neigh(i) bestnode(1) dtogo+dcur dcur];
+                   end
+                end
+              
+            case 2 %Breadth-First   (QUEUE)
+                % If neighbour is not in open set, add it
+                if (isempty(found))
+                   OpenSet = [OpenSet; neigh(i) bestnode(1) dcur dcur];     
+        
+                else % If in open set, update cost if better    
+                   if (dcur < OpenSet(found,4))
+                      OpenSet(found,:) = [neigh(i) bestnode(1) dcur dcur];
+                   end
+                end                   
+            
+            case 3 %Depth-First  (STACK)
+                % If neighbour is not in open set, add it
+                if (isempty(found))
+                   OpenSet = [neigh(i) bestnode(1) dcur dcur; OpenSet];
+                else % If in open set, update cost if better    
+                   if (dcur < OpenSet(found,4))
+                      OpenSet(found,:) = [neigh(i) bestnode(1) dcur dcur];
+                   end
+                end                   
+
+            case 4 %Dijkstra's 
+                % If neighbour is not in open set, add it
+                if (isempty(found)) 
+                   OpenSet = [OpenSet; neigh(i) bestnode(1) dcur dcur]; 
+                
+                else % If neighbour is in open set, check if new route is better
+                    if (dcur < OpenSet(found,4))
+                       OpenSet(found,:) = [neigh(i) bestnode(1) dcur dcur];
+                    end
+                end                    
+            end     
+        
+        
+    end
+    
+    switch method
+        case 1 % Astart
+               OpenSet = OpenSet([1:best-1 best+1:end],:); % remove best node from open set                            
+        case 4 %Dijkstra's     
+               OpenSet = OpenSet([1:best-1 best+1:end],:); % remove best node from open set                
+    end
+    
+    % Plot active nodes for this step
+    figure(1);hold on;
+    plot(nodes(C(:,1),1),nodes(C(:,1),2), 'ko','MarkerSize',6,'LineWidth',2);
+    plot(nodes(bestnode(1),1),nodes(bestnode(1),2), 'go','MarkerSize',6,'LineWidth',2);
+    for i=1:length(neigh)
+        plot(nodes(neigh(i),1),nodes(neigh(i),2), 'mo');
+        plot([nodes(bestnode(1),1) nodes(neigh(i),1)],[nodes(bestnode(1),2) nodes(neigh(i),2)], 'm');
+    end
+    if ((createVideo >0) && (createVideo <=1))
+        writeVideo(vidObj, getframe(gcf));
     end
-    O = O([1:best-1 best+1:end],:);
 end
 
+%%======================================================== END OF MAIN
 % Find final path through back tracing
```

Comment:
- It would be nice to show a table of how many nodes were expanded and the run time of each algorithm for comparison
- I think a better visualization might be a subplot showing the final solution of all the 4 plots together.
- The Plots could also include the title to state which method is being used